File: solutions/python/trees/find-all-the-lonely-nodes.py
# LeetCode 1469 — Find All The Lonely Nodes (Easy)
# Category: Trees · Approach: DFS
# Time: O(n) | Space: O(h)
# Source: https://leetcode.com/problems/find-all-the-lonely-nodes/

def get_lonely_nodes(root: "TreeNode") -> list[int]:
    out: list[int] = []
    def dfs(node):
        if not node:
            return
        if node.left and not node.right:
            out.append(node.left.val)
        if node.right and not node.left:
            out.append(node.right.val)
        dfs(node.left)
        dfs(node.right)
    dfs(root)
    return out

File: solutions/typescript/trees/find-all-the-lonely-nodes.ts
// LeetCode 1469 — Find All The Lonely Nodes (Easy)
// Category: Trees · Approach: DFS
// Time: O(n) | Space: O(h)
// Source: https://leetcode.com/problems/find-all-the-lonely-nodes/

function getLonelyNodes(root: TreeNode | null): number[] {
  const out: number[] = [];
  const dfs = (n: TreeNode | null): void => {
    if (!n) return;
    if (n.left && !n.right) out.push(n.left.val);
    if (n.right && !n.left) out.push(n.right.val);
    dfs(n.left);
    dfs(n.right);
  };
  dfs(root);
  return out;
}
